### PR TITLE
refactor(lane change): lane change class rework

### PR DESCRIPTION
**Pull request by zulfaqar-azmi-t4**
## Description

refactor lane change class so that it is more resemble to scene module manager

⬇️ This launcher needs to be merge first
https://github.com/autowarefoundation/autoware_launch/pull/248

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Review by satoshi-ota — COMMENTED** [review]
I think we have to update this line, `lane_change_right || lane_change_left`.

https://github.com/autowarefoundation/autoware.universe/blob/0d4e4444fdb1ea039b26d7e87b3c4cd9b130c162/planning/behavior_path_planner/src/planner_manager.cpp#L284

File: planning/behavior_path_planner/src/planner_manager.cpp [hunk 1]
```diff
@@ -281,7 +281,7 @@ BehaviorModuleOutput PlannerManager::update(const std::shared_ptr<PlannerData> &
     if ((*itr)->getCurrentStatus() != ModuleStatus::RUNNING) {
       if (itr == approved_module_ptrs_.begin()) {
         // update root lanelet when the lane change is done.
-        if (name == "lane_change") {
+        if (name.find("lane_change") != std::string::npos) {
```
Comment (satoshi-ota) [inline planning/behavior_path_planner/src/planner_manager.cpp:284]: Great :+1:

File: planning/route_handler/include/route_handler/route_handler.hpp [hunk 2]
```diff
@@ -259,9 +259,19 @@ class RouteHandler
     const lanelet::ConstLanelet & lanelet, const double length,
     const lanelet::ConstLanelets & exclude_lanelets = {}) const;
 
-  int getNumLaneToPreferredLane(const lanelet::ConstLanelet & lanelet) const;
+  /**
+   * Query input lanelet  to see whether it exist in the preffered lane. If it doesn't exist, return
```
Comment (satoshi-ota) [inline planning/route_handler/include/route_handler/route_handler.hpp:263]:
```suggestion
   * Query input lanelet  to see whether it exist in the preferred lane. If it doesn't exist, return
```

File: planning/route_handler/include/route_handler/route_handler.hpp [hunk 3]
```diff
@@ -259,9 +259,19 @@ class RouteHandler
...
+   * the number of lane-changeable lane to the preferred lane.
+   * @param Desired lanelet to query
+   * @param lane change direction
+   * @return number of lanes from input to the preffered lane
```
Comment (satoshi-ota) [inline planning/route_handler/include/route_handler/route_handler.hpp:267]:
```suggestion
   * @return number of lanes from input to the preferred lane
```

**Comment by zulfaqar-azmi-t4** [comment]
Evaluator test seems okay
https://github.com/tier4/pilot-auto/commit/cc281caf6a6d0dc7a95f6bd92aa5e95a84984f1e

**Comment by satoshi-ota** [comment]
@zulfaqar-azmi-t4 LGTM :+1:  I'll approve it after you resolve conflicts.

**Review by satoshi-ota — APPROVED** [review]

File: planning/behavior_path_planner/include/behavior_path_planner/util/lane_change/lane_change_module_data.hpp [hunk 4]
```diff
@@ -79,6 +79,11 @@ struct LaneChangeTargetObjectIndices
   std::vector<size_t> target_lane{};
   std::vector<size_t> other_lane{};
 };
+
+enum class LaneChangeType {
+  Normal = 0,
+  ExternalRequest = 0,
```
Comment (rej55) [inline planning/behavior_path_planner/include/behavior_path_planner/util/lane_change/lane_change_module_data.hpp:85]: Is this suggestion right?

```suggestion
  ExternalRequest = 1,
```

**Review by rej55 — APPROVED** [review]

**Pull request merged**